### PR TITLE
NXP-22163: basic service for attach methods

### PR DESCRIPTION
**Pull request by mcedica**

### CODE REVIEW

File: nuxeo-retention-core/pom.xml [hunk 1]
```diff
@@ -12,4 +12,31 @@
...
+      <groupId>org.nuxeo.ecm.core</groupId>
+      <artifactId>nuxeo-core-query</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.nuxeo.ecm.platform</groupId>
+      <artifactId>nuxeo-platform-query-api</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.nuxeo.runtime</groupId>
+      <artifactId>nuxeo-runtime-test</artifactId>
```
Comment (akervern) [inline nuxeo-retention-core/pom.xml]: scope test?

File: nuxeo-retention-core/pom.xml [hunk 2]
```diff
@@ -12,4 +12,31 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.nuxeo.ecm.core</groupId>
+      <artifactId>nuxeo-core-test</artifactId>
```
Comment (akervern) [inline nuxeo-retention-core/pom.xml]: scope test too.

File: nuxeo-retention-core/src/main/java/org/nuxeo/ecm/retention/service/RetentionService.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * Attaches the given rule to the document using an unrestricted session. 
+     * 
+     * @param ruleId
+     * @param doc
+     * @param session
+     * @since 9.2
+     */
+    void attachRule(String ruleId, DocumentModel doc, CoreSession session);
```
Comment (akervern) [inline nuxeo-retention-core/src/main/java/org/nuxeo/ecm/retention/service/RetentionService.java]: I would use the doc#getSession as base Session. And throw an exception for detached doc for now.

File: nuxeo-retention-core/src/main/java/org/nuxeo/ecm/retention/service/RetentionComponent.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            // document later
+            RetentionRecordUpdateWork work = new RetentionRecordUpdateWork(ruleId);
+            work.setDocuments(session.getRepositoryName(), docIds);
+            workManager.schedule(work, WorkManager.Scheduling.IF_NOT_SCHEDULED, true);
+            offset += maxResult;
+        } while (nextDocumentsToBeUpdated.size() == maxResult && pp.isNextPageAvailable());
+    }
+
+    final class CoreQueryDocumentPageProviderDescriptor extends GenericPageProviderDescriptor {
+        private static final long serialVersionUID = 1L;
```
Comment (akervern) [inline nuxeo-retention-core/src/main/java/org/nuxeo/ecm/retention/service/RetentionComponent.java]: Why don't you use org.nuxeo.ecm.platform.query.core.CoreQueryPageProviderDescriptor?

File: nuxeo-retention-core/src/main/java/org/nuxeo/ecm/retention/service/RetentionService.java (hunk 3 above)
Comment (mcedica) [inline nuxeo-retention-core/src/main/java/org/nuxeo/ecm/retention/service/RetentionService.java]: OK, but why? :)

Comment (akervern) [inline nuxeo-retention-core/src/main/java/org/nuxeo/ecm/retention/service/RetentionService.java]: cause it's useless to have a parameter that is already a field of another parameter.

**Pull request merged**